packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,455 @@
+    await removeExpiredFields(devProfile, devUserData);
+    await ensureActiveEnv(projectSettings);
+  }
+}
+async function moveIconsToResourceFolder(templateAppPackage: string): Promise<void> {
+  // see AppStudioPluginImpl.buildTeamsAppPackage()
+  const manifest: TeamsAppManifest = await readJson(
+    path.join(templateAppPackage, "manifest.template.json")
+  );
+  const hasColorIcon = manifest.icons.color && !manifest.icons.color.startsWith("https://");
+  const hasOutlineIcon = manifest.icons.outline && !manifest.icons.outline.startsWith("https://");
+  if (!hasColorIcon || !hasOutlineIcon) {
+    return;
+  }
+  // move to resources
+  const resource = path.join(templateAppPackage, "resources");
+  await fs.ensureDir(resource);
+  await fs.move(
+    path.join(templateAppPackage, manifest.icons.color),
+    path.join(resource, manifest.icons.color)
+  );
+
+  await fs.move(
+    path.join(templateAppPackage, manifest.icons.outline),
+    path.join(resource, manifest.icons.outline)
+  );
+  // update icons
+  manifest.icons.color = `resources/${manifest.icons.color}`;
+  manifest.icons.outline = `resources/${manifest.icons.outline}`;
+  await fs.writeFile(
+    path.join(templateAppPackage, "manifest.template.json"),
+    JSON.stringify(manifest, null, 4)
+  );
+}
+
+async function removeExpiredFields(devProfile: string, devUserData: string): Promise<void> {
+  // remove fx-resource-local-debug.trustDevCert, solution.programmingLanguage, fx-resource-function.defaultFunctionName
+  const profileData = await readJson(devProfile);
+  if (profileData[PluginNames.LDEBUG]) {
+    delete profileData[PluginNames.LDEBUG];
+  }
+  if (profileData[PluginNames.SOLUTION] && profileData[PluginNames.SOLUTION][programmingLanguage]) {
+    delete profileData[PluginNames.SOLUTION][programmingLanguage];
+  }
+  if (profileData[PluginNames.FUNC] && profileData[PluginNames.FUNC][defaultFunctionName]) {
+    delete profileData[PluginNames.FUNC][defaultFunctionName];
+  }
+  await fs.writeFile(devProfile, JSON.stringify(profileData, null, 4), { encoding: "UTF-8" });
+  const trustDevCertKey = `${PluginNames.LDEBUG}.trustDevCert`;
+  const secrets: Record<string, string> = deserializeDict(await fs.readFile(devUserData, "UTF-8"));
+
+  if (secrets[trustDevCertKey]) {
+    delete secrets[trustDevCertKey];
+    await fs.writeFile(devUserData, serializeDict(secrets), { encoding: "UTF-8" });
+  }
+}
+
+function getConfigDevJson(appName: string): EnvConfig {
+  const envConfig: EnvConfig = {
+    $schema:
+      "https://raw.githubusercontent.com/OfficeDev/TeamsFx/dev/packages/api/src/schemas/envConfig.json",
+    manifest: {
+      description: `You can customize the 'values' object to customize Teams app manifest for different environments. Visit https://aka.ms/teamsfx-config to learn more about this.`,
+      values: {
+        appName: {
+          short: appName,
+          full: `Full name for ${appName}`,
+        },
+      },
+    },
+  };
+  return envConfig;
```

Comment:
how about using `environmentManager.newEnvConfigData()` to initialize the env config to ensure one source of truth?

Reply:
Good idea, will add.